File: RxTango/python/examples/sliding_average.py
"""Sliding average over a rolling window using buffer_with_count.

Polls double_scalar and computes a rolling mean of the last N samples.
No circular buffer, no index arithmetic — one operator.

Mirrors `TangoTestSlidingAverage.java` (`buffer(N, 1)` → `buffer_with_count(N, 1)`).

Usage:
    python sliding_average.py [device] [window] [interval-ms]

    device       defaults to tango://localhost:10000/sys/tg_test/1
    window       samples in the rolling window; defaults to 5
    interval-ms  poll interval; defaults to 500
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import read_attribute


async def main() -> None:
    device      = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    window      = int(sys.argv[2]) if len(sys.argv) > 2 else 5
    interval_ms = int(sys.argv[3]) if len(sys.argv) > 3 else 500

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(
        f"Sliding average (N={window}) on {device}  "
        f"every {interval_ms} ms  (Ctrl+C to stop)\n"
    )
    print(f"  {'raw':>14}  {'avg(N={window})':>14}  {'delta':>14}")
    print("  " + "-" * 46)

    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_attribute(device, "double_scalar")),
        ops.buffer_with_count(count=window, skip=1),
        ops.map(lambda buf: (buf[-1], sum(buf) / len(buf))),
    ).subscribe(
        on_next=lambda pair: print(
            f"  {pair[0]:>+14.6f}  {pair[1]:>14.6f}  {pair[0] - pair[1]:>+14.6f}"
        ),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/sliding_average.cpp
/**
 * Sliding (rolling) average using buffer_with_time / window operator.
 *
 * Uses rxcpp's buffer(N, skip=1) to create overlapping windows, then maps
 * each window to its mean.  Identical to the Python sliding-window pattern:
 *   ops.buffer_with_count(N, 1) → ops.map(mean)
 *
 * Mirrors Python's sliding_average.py and Java's TangoTestSlidingAverage.java.
 *
 * Usage:
 *   ./sliding_average [device] [attribute] [window-size] [interval-ms]
 *   defaults: sys/tg_test/1  double_scalar  5  200
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device      = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr        = argc > 2 ? argv[2] : "double_scalar";
    const int         window_size = argc > 3 ? std::stoi(argv[3]) : 5;
    const int         interval_ms = argc > 4 ? std::stoi(argv[4]) : 200;

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Sliding average (window=" << window_size << ") of "
              << device << "/" << attr << "  (Ctrl+C to stop)\n\n";

    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([device, attr](long) {
            return rxtango::read_attribute<double>(device, attr);
        })
        // Overlapping windows of size N, advancing by 1 each step
        .buffer(window_size, 1)
        // Skip partial leading windows (size < window_size)
        .filter([window_size](const std::vector<double>& w) {
            return static_cast<int>(w.size()) == window_size;
        })
        .map([](const std::vector<double>& w) {
            return std::accumulate(w.begin(), w.end(), 0.0) / w.size();
        })
        .subscribe(
            [](double avg) { std::cout << "  avg: " << avg << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
